login/service.go

```diff
@@ -0,0 +1,163 @@
+	"fmt"
+
+	"github.com/almighty/almighty-core/account"
+	"github.com/almighty/almighty-core/app"
+	"github.com/almighty/almighty-core/token"
+	uuid "github.com/satori/go.uuid"
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+)
+
+// Service defines the basic entrypoint required to perform a remote oauth login
+type Service interface {
+	Perform(ctx *app.AuthorizeLoginContext) error
+}
+
+// NewGitHubOAuth creates a new login.Service capable of using GitHub for authorization
+func NewGitHubOAuth(config *oauth2.Config, identities account.IdentityRepository, users account.UserRepository, tokenManager token.Manager) Service {
+	return &gitHubOAuth{
+		config:       config,
+		identities:   identities,
+		users:        users,
+		tokenManager: tokenManager,
+	}
+}
+
+type gitHubOAuth struct {
+	config       *oauth2.Config
+	identities   account.IdentityRepository
+	users        account.UserRepository
+	tokenManager token.Manager
+}
+
+// TEMP: This will leak memory in the long run with many 'failed' redirect attemts
+var stateReferer = map[string]string{}
+
+func (gh *gitHubOAuth) Perform(ctx *app.AuthorizeLoginContext) error {
+	state := ctx.Params.Get("state")
+	code := ctx.Params.Get("code")
+	if code != "" {
+		// After redirect from oauth provider
+
+		// validate known state
+		var referer string
+		defer func() {
+			delete(stateReferer, state)
+		}()
+
+		if referer = stateReferer[state]; referer == "" || state == "" {
+			return ctx.Unauthorized()
+		}
+
+		ghtoken, err := gh.config.Exchange(ctx, code)
+		if err != nil {
+			fmt.Println(err)
+			return ctx.Unauthorized()
+		}
+
+		emails, err := gh.getUserEmails(ctx, ghtoken)
+		fmt.Println(emails)
+
+		primaryEmail := filterPrimaryEmail(emails)
+		if primaryEmail == "" {
+			fmt.Println("No primary email found?! ", emails)
+			return ctx.Unauthorized()
+		}
+		users, err := gh.users.Query(account.UserByEmails([]string{primaryEmail}), account.UserWithIdentity())
+		if err != nil {
+			fmt.Println(err)
+			return ctx.Unauthorized()
+		}
+		var identity account.Identity
+		if len(users) == 0 {
+			// No User found, create new User and Identity
+			ghUser, err := gh.getUser(ctx, ghtoken)
+			if err != nil {
+				fmt.Println(err)
+				return ctx.Unauthorized()
+			}
+			fmt.Println(ghUser)
+
+			identity := account.Identity{
+				FullName: ghUser.Name,
+				ImageURL: ghUser.AvatarURL,
+			}
+			gh.identities.Create(ctx, &identity)
+			gh.users.Create(ctx, &account.User{Email: primaryEmail, Identity: identity})
+		} else {
+			identity = users[0].Identity
+		}
+
+		fmt.Println("Identity: ", identity)
+
+		// register other emails in User table?
+
+		// generate token
+		almtoken := gh.tokenManager.Generate(identity)
+
+		ctx.ResponseData.Header().Set("Location", referer+"?token="+almtoken)
+		return ctx.TemporaryRedirect()
+	}
+
+	// First time access, redirect to oauth provider
+
+	// store referer id to state for redirect later
+	referer := ctx.RequestData.Header.Get("Referer")
```

Comment:
Referer is a standard HTTP Header that is 'normally' set on all requests by the browser.